### PR TITLE
RSDK-5278: Add context logger methods.

### PR DESCRIPTION
**Pull request by dgottlieb**

### CODE REVIEW

File: logging/impl.go [hunk 1]
```diff
@@ -142,33 +143,48 @@ func (imp *impl) log(entry *LogEntry) {
 	}
 }
 
-// Constructs the log message by forwarding to `fmt.Sprint`.
-func (imp *impl) format(logLevel Level, args ...interface{}) *LogEntry {
+// emptyTraceKey is a sentinal value for internal helper functions when there's no context in debug
+// mode.
+const emptyTraceKey = ""
```
Comment (benjirewis) [inline logging/impl.go:148]: [opt-nit] `GetName` could probably return `emptyTraceKey` in the case of no trace key to be clear.

File: logging/impl.go [hunk 2]
```diff
@@ -192,89 +208,197 @@ func (imp *impl) formatw(logLevel Level, msg string, keysAndValues ...interface{
...
-		imp.log(imp.formatw(DEBUG, msg, keysAndValues...))
+		imp.log(imp.formatw(DEBUG, emptyTraceKey, msg, keysAndValues...))
+	}
+}
+
+func (imp *impl) CDebugw(ctx context.Context, msg string, keysAndValues ...interface{}) {
+	dbgName := GetName(ctx)
+
+	// We log if the logger is configured for debug, or if there's a trace key.
+	if imp.shouldLog(DEBUG) || dbgName != emptyTraceKey {
```
Comment (benjirewis) [inline logging/impl.go:249]: [q] So if we're on info-level logging, we would still log a debug-level log that contained a trace key? That makes sense to me, but just want to verify.

File: logging/impl_test.go [hunk 3]
```diff
@@ -190,3 +191,29 @@ func TestConsoleOutputFormat(t *testing.T) {
...
+	ctx = EnableDebugModeWithKey(ctx, traceKey)
+	logger.CDebug(ctx, "Debug log")
+	assertLogMatches(t, notStdout,
+		`2023-10-30T09:12:09.459Z	DEBUG	impl	logging/impl_test.go:200	Debug log	{"traceKey":"foobar"}`)
+
+	logger.CDebugf(ctx, "Debugf log %v", "debugf")
+	assertLogMatches(t, notStdout,
+		`2023-10-30T09:12:09.459Z	DEBUG	impl	logging/impl_test.go:200	Debugf log debugf	{"traceKey":"foobar"}`)
+
+	logger.CDebugw(ctx, "Debugw log", "key", "value")
```
Comment (benjirewis) [inline logging/impl_test.go:216]: [nit] Could also test info, warn and error levels.

**Review by benjirewis — COMMENTED** [review]
Nice! Just some nits and qs.

File: logging/impl.go (hunk 2 above)
Comment (dgottlieb) [reply logging/impl.go:245]: Correct. The log level outputted is 1-1 with the log function called.

File: logging/impl_test.go (hunk 3 above)
Comment (dgottlieb) [reply logging/impl_test.go]: We could :P

Done

File: logging/impl.go (hunk 1 above)
Comment (dgottlieb) [reply logging/impl.go]: Done

**Review by benjirewis — APPROVED** [review]
LGTM thanks! 🧑‍🔧

File: logging/impl.go (hunk 2 above)
Comment (dgottlieb) [reply logging/impl.go:245]: Offline clarification:
If the log "filter" is at INFO, a "debug context" logs at the level of the written log line. E.g:
```
logger.SetLevel(INFO)
logger.Debug("won't output")
logger.CDebug(EnableDebugMode(ctx), "will output at debug")
```
Gives a log line such as:
```
2023-10-30T09:12:09.459Z	DEBUG	impl	logging/impl_test.go:200	will output at debug	{"traceKey":"foobar"}
```

File: logging/impl.go [hunk 4]
```diff
@@ -192,89 +204,197 @@ func (imp *impl) formatw(logLevel Level, msg string, keysAndValues ...interface{
...
-		imp.log(imp.format(ERROR, args...))
+		imp.log(imp.format(ERROR, emptyTraceKey, args...))
+	}
+}
+
+func (imp *impl) CError(ctx context.Context, args ...interface{}) {
+	dbgName := GetName(ctx)
+
+	// We log if the logger is configured for error, or if there's a trace key.
+	if imp.shouldLog(WARN) || dbgName != emptyTraceKey {
```
Comment (cheukt) [inline logging/impl.go:350]: shouldLog(ERROR)?

File: logging/impl.go [hunk 5]
```diff
@@ -192,89 +204,197 @@ func (imp *impl) formatw(logLevel Level, msg string, keysAndValues ...interface{
...
-		imp.log(imp.formatf(ERROR, template, args...))
+		imp.log(imp.formatf(ERROR, emptyTraceKey, template, args...))
+	}
+}
+
+func (imp *impl) CErrorf(ctx context.Context, template string, args ...interface{}) {
+	dbgName := GetName(ctx)
+
+	// We log if the logger is configured for error, or if there's a trace key.
+	if imp.shouldLog(WARN) || dbgName != emptyTraceKey {
```
Comment (cheukt) [inline logging/impl.go:365]: shouldLog(ERROR)?

File: logging/impl.go [hunk 6]
```diff
@@ -192,89 +204,197 @@ func (imp *impl) formatw(logLevel Level, msg string, keysAndValues ...interface{
...
-		imp.log(imp.formatw(ERROR, msg, keysAndValues...))
+		imp.log(imp.formatw(ERROR, emptyTraceKey, msg, keysAndValues...))
+	}
+}
+
+func (imp *impl) CErrorw(ctx context.Context, msg string, keysAndValues ...interface{}) {
+	dbgName := GetName(ctx)
+
+	// We log if the logger is configured for error, or if there's a trace key.
+	if imp.shouldLog(WARN) || dbgName != emptyTraceKey {
```
Comment (cheukt) [inline logging/impl.go:380]: shouldLog(ERROR)?

**Review by cheukt — APPROVED** [review]
LGTM % nits

**Pull request merged**